### DISCUSSION TITLE
Pydantic V2 Custom data type's JSON schema type is not used as a path parameter when the JSON type and real type is different

### DISCUSSION THREAD

**Discussion by Greg-NetDuma** [fastapi/fastapi#10219] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Dummy functions if you just copy-paste this to test
def convertFromSemVer(version: str) -> int:
    """Converts a semver version string to an integer, so it can be compared with other versions efficiently."""

    return 1
def convertToSemVer(version: int) -> str:
    """Converts an integer to a semver version string.

    Args:
        version (int): The integer to convert.
    """

    return str(version)


FirmwareVersion = Annotated[
    int,
    BeforeValidator(lambda x: convertFromSemVer(x) if isinstance(x, str) else x),
    PlainSerializer(lambda x: convertToSemVer(x) if isinstance(x, int) else x, return_type=str),
    WithJsonSchema({
        'type': 'string',
        'examples': ['3.0.0', '3.1.2-pre+26515115'],
    }),
]
''' Integer representation of SemVer Firmware Version. You can use this type to validate and serialize/deserialize SemVer Firmware Versions.

You can manually convert between `str` and `int` representation using `convertFromSemVer` and `convertToSemVer` functions.
'''

@router.get("/{min_version}", response_model=FirmwareVersion)
async def test(
        min_version: FirmwareVersion,
    ) -> FirmwareVersion:

    return min_version
```

### Description

`FirmwareVersion` should be treated as a string in docs, during validation and serialization, but it's not.

* Open the browser and try to call the endpoint
* Validation fails, because FirmwareVersion is treated as an integer instead of a string if it's in the path.

Note this Custom Data Type works correctly inside `BaseModel` and `Annotated[FirmwareVersion, Query()/Form()]`

### Operating System

Linux

### FastAPI Version

0.103.1

### Pydantic Version

2.3.0

### Python Version

3.11

### Additional Context

Screenshot of the documentation produced, no both `min_version`s are the same type:

![image](https://github.com/tiangolo/fastapi/assets/69147407/a0b5bea0-e71e-4653-b359-c5af7ae10a1b)

**Comment by YuriiMotov** [comment]
Fixed.
Works well in current version (fastapi-0.115.12, pydantic-2.11.5)
